# Suicide Prediction using Machine Learning.

Here is the outline of the procedure:
1. Question
2. Understand the data
3. Manipulate the data
4. Fit & test models
5. Deploy models

## 1. Question

Due to the random nature of suicidality, traditional statistical methods have reported poor performance (SOURCE). This can be attributed to the notion that no single risk factor or risk assessment approach can be directly applied to suicide prediction with great accuracy. This makes relying on few strong predictors implausible and places greater emphasis on examining a larger number of factors. This requirement of complex combinations of hundreds of factors, is not ideal for traditional statistical methods. Furthermore, traditional statistical methods require the ideal algorithm to be designed before the collection of the data (a priori) to account for certain assumptions. In contrast, DL algorithms are self-optimizing and increase their predictive performance automatically through a method called back-propagation. This use of optimization means that the DL algorithm instead, learns from the data itself giving the analyst data driven insights. This also allows DL algorithms to harness high-dimensional data where the large number of predictors allow it to learn the appropriate parameters which can be applied to future datasets. Finally, a meta-analysis by Franklin et al., (2017) found that traditional statistical methods in suicide attempts have only produced models with near chance level predictive performance, highlighting a need for more advanced data driven methods. 

Emerging DL research in suicidality has shown promising results compared to traditional methods. Hack et al. (2017) examined suicidality in the Grady Trauma project for those looking for primary care in large urban hospitals. Their dataset comprised of 1017 participants, split into training and test sets (80/20) both of which had stratified 16% suicide attempters. They deployed Support Vector Machines (SVMs) and least absolute shrinkage and selection operator (LASSO) models with 100 iterations of which scored an area under the curve (AUC) of 70% and 71% respectively. Walsh et al (2017) used machine learning to predict suicide attempts from electronic health records. They developed a Random Forest algorithm that utilises an ensemble of decision trees and was run for 500 iterations. Their model accurately predicted future suicide attempts with an AUC of 0.84 and found that accuracy improved from 720 days to 7 days before the suicide attempt.  

Walsh et al., (2017) found that “recurrent depression with psychosis, schizophrenia, and schizoaffective disorder were consistently ranked highly in importance. Age and diagnoses of dependence on opioids, sedative-hypnotics, and cannabis increased in relative importance as prediction windows shortened. Some codes that likely indicate prior suicide attempts were also consistently predictive: poisoning, the most common mechanism of prior nonfatal suicide attempts in these data; injuries by firearms; and injuries “NEC” or not elsewhere classifiable. Medication classes such as selective serotonin reuptake inhibitors (SSRIs), benzodiazepines, anilides (such as acetaminophen), and propionic acid derivatives (such as ibuprofen) appear stronger within longer prediction windows. Melatonin receptor agonists such as melatonin supplements gain relative importance closer to the suicide attempt (i.e., shorter prediction windows).”

## 2. Understand the data

In [8]:
import numpy as np 
import pandas as pd 
baseline = pd.read_csv("../../data/t1_beta.csv")

In [9]:
baseline.head()

Participant_ID  wave       today      INT_ID      phone  Postcode  \
0           26885     0   6/25/2013          TM  Telephone      4500   
1           27928     0  10/25/2013  Don't Know  Telephone      3020   
2           27112     0   8/23/2013       Other  Telephone      5052   
3           26719     0   6/20/2013          CO  Telephone      5276   
4           28473     0    1/7/2014  Don't Know  Telephone      4575   

         dob  doby  dobm  dobd  ...  OME_over90 threeorMoreCC fourorMoreCC  \
0  10/8/1951  1951    10     8  ...           0             0            0   
1  3/26/1974  1974     3    26  ...           0             0            0   
2  11/8/1960  1960    11     8  ...           0             1            0   
3  7/15/1950  1950     7    15  ...           0             0            0   
4  7/30/1948  1948     7    30  ...           0             1            1   

  Deceased1 Withdrew_self Withdrawn_all Cannabis_12m Cannabis12m_notbaseline  \
0        No                           0           No                      No   
1        No                           0           No                      No   
2        No           Yes             1           No                      No   
3       Yes                           2           No                      No   
4        No                           0           No                      No   

  Cannabisbaseline_not12m Cannabisbaseline_12m  
0                      No                   No  
1                      No                   No  
2                      No                   No  
3                      No                   No  
4                      No                   No  

[5 rows x 2903 columns]

In [10]:
print("Shape:", baseline.shape)

Shape: (1514, 2903)


We can see that there are 1514 rows with 2903 variables for the baseline data.

In [3]:
# Variables we want are saved in a txt file in the current directory 
!ls

Untitled.ipynb  model1_vars.txt


In [56]:
filename = "model1_vars.txt"

# using the with construct closes the file automatically when finished with it
with open(filename, "r") as file:
    variables = file.readlines()
variables = [i.strip() for i in variables]

# Now we subset the data based on the variables we are intersted in. 
model1_data = df[variables]

In [76]:
# Ideation
x = model1_data.drop('Suicidal_Thoughts_12m', axis=1)
x =x.replace(" ", np.NaN)

y = model1_data[['Suicidal_Thoughts_12m']]

# Attempts
# x = model1_data.drop('Suicide_Attempts_12m', axis=1)
# y = model1_data[['Suicide_Attempts_12m']]

In [86]:
to_drop1 = variables[21:32]
to_drop2 = variables[38:51]

to_drop = to_drop1 + to_drop2

['Suicide_Plan_Mthd_12m',
 'Suicide_Plan_gun',
 'Suicide_Plan_carb_monox',
 'Suicide_Plan_cut_stab',
 'Suicide_Plan_pills',
 'Suicide_Plan_jump_height',
 'Suicide_Plan_jump_train',
 'Suicide_Plan_hang',
 'Suicide_Plan_od_heroin',
 'Suicide_Plan_od_op',
 'Suicide_Plan_choke',
 'Suicide_Attempt_gun',
 'Suicide_Attempt_carb_monox',
 'Suicide_Attempt_cut_stab',
 'Suicide_Attempt_pills',
 'Suicide_Attempt_jump_hght',
 'Suicide_Attempt_jump_train',
 'Suicide_Attempt_hang',
 'Suicide_Attempt_od_heroin',
 'Suicide_Attempt_od_op',
 'Suicide_Attempt_choke',
 'Suicide_Attempt_drown',
 'Suicide_Attempt_other',
 'Suicide_Attempt_other_sp']

# 3. Fit the model(s)

Here we will be testing many models:
1. Logistic regression
2. Support Vector Machines
3. Tree models
4. Artificial Neural Networks

### 3.1. Logistic Regression

### 3.2. Support Vector Machine

### 3.3. Tree Models

### 3.4. Artificial Neural Network

**REFERENCES**

Franklin, J., Riberio, J., Fox, K., Bentley K, Kleiman, E., Huang, X., & Nock, M. (2017). 			     	Psychological Bulletin , 187-232.

Hack, L., Jovanovic, T., Carter, S., Ressler, K., & Smith, A. (2017). Suicide prediction using machine 	   	learning techniques in screening clinician derived data. Biological Psychiatry, 361-361.

Walsh, C., Ribeiro, J., & Franklin, J. (2017). Predicting Risk of Suicide Attempts Over Time Through   	Machine Learning. Clinical Psychological Science, 457-469.

# End.